flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/DefaultAllocatedSlotPool.java

```diff
@@ -218,11 +205,59 @@ public Collection<FreeSlotInfo> getFreeSlotsInformation() {
         return freeSlotInfos;
     }
 
+    public double getTaskExecutorUtilization(ResourceID resourceId) {
+        Set<AllocationID> slots = slotsPerTaskExecutor.get(resourceId);
+        Preconditions.checkNotNull(slots, "There is no slots on %s", resourceId);
+
+        return (double) (slots.size() - freeSlots.getFreeSlotsNumberOfTaskExecutor(resourceId))
+                / slots.size();
+    }
+
     @Override
     public Collection<? extends SlotInfo> getAllSlotsInformation() {
         return registeredSlots.values();
     }
 
+    private static final class FreeSlots {
+        /** Map containing all free slots and since when they are free. */
+        private final Map<AllocationID, Long> freeSlotsSince = new HashMap<>();
+
+        /** Index containing a mapping between TaskExecutors and their free slots number. */
+        private final Map<ResourceID, Integer> freeSlotsNumberPerTaskExecutor = new HashMap<>();
+
+        public void addFreeSlot(AllocatedSlot slot, long currentTime) {
+            Preconditions.checkState(
+                    !freeSlotsSince.containsKey(slot.getAllocationId()),
+                    "Slot with id %s has been freed",
+                    slot.getAllocationId());
+            freeSlotsSince.put(slot.getAllocationId(), currentTime);
+            freeSlotsNumberPerTaskExecutor.put(
+                    slot.getTaskManagerId(),
+                    freeSlotsNumberPerTaskExecutor.getOrDefault(slot.getTaskManagerId(), 0) + 1);
+        }
+
+        public Long removeFreeSlot(AllocatedSlot slot) {
```

Comment:
I suggest to take advantage of the return value of the `remove` method , and use `compute` mothod here, a suggestion:
```
Long freeSince = freeSlotsSince.remove(slot.getAllocationId());
if (freeSince != null) {
    freeSlotsNumberPerTaskExecutor.computeIfPresent(
            slot.getTaskManagerId(),
            (ignore, count) -> {
                int newCount = count - 1;
                return newCount == 0 ? null : newCount;
            });
}

return freeSince;
```
Please correct me if I'm wrong :)